ultravox/data/datasets.py

```diff
@@ -266,7 +267,32 @@ def _get_messages(
     return messages
 
 
-class VoiceDataset(abc.ABC, data.IterableDataset):
+class SizedIterableDataset(data.IterableDataset):
+    """
+    A wrapper for a Dataset that provides a length method and verifies the count.
+    """
+
+    def __init__(self, dataset: data.Dataset, estimated_length: int) -> None:
+        self._dataset = dataset
+        self._estimated_length = estimated_length
+
+    def __iter__(self):
+        self._actual_length = 0
+        for item in self._dataset:
+            self._actual_length += 1
+            yield item
+
+        if self._actual_length != self._estimated_length:
+            warnings.warn(
+                f"Mismatch between estimated length ({self._estimated_length}) "
+                f"and actual length ({self._actual_length}) for dataset of type {type(self._dataset)}"
+            )
+
```

Comment:
If we add the previous warning, we should only trigger the warning if `self._actual_length < self._estimated_length`